prow/repoowners/repoowners.go

```diff
@@ -215,36 +221,58 @@ func (c *Client) LoadRepoOwners(org, repo, base string) (RepoOwner, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	entry, ok := c.cache[fullName]
-	if !ok || entry.sha != sha || entry.owners == nil || entry.owners.enableMDYAML != mdYaml {
+	if !ok || entry.sha != sha || !entry.isValidOwners(mdYaml) {
 		gitRepo, err := c.git.Clone(cloneRef)
 		if err != nil {
 			return nil, fmt.Errorf("failed to clone %s: %v", cloneRef, err)
 		}
 		defer gitRepo.Clean()
-		if err := gitRepo.Checkout(base); err != nil {
-			return nil, err
-		}
 
-		if entry.aliases == nil || entry.sha != sha {
-			// aliases must be loaded
-			entry.aliases = loadAliasesFrom(gitRepo.Dir, log)
+		reusable := entry.isValid(mdYaml)
+		// In most sha changed cases, the files associated with the owners are unchanged.
+		// The cached entry can continue to be used, so need do git diff
+		if reusable {
+			changes, err := gitRepo.Diff(sha, entry.sha)
+			if err != nil {
+				return nil, fmt.Errorf("failed to diff %s with %s", sha, entry.sha)
+			}
+			for _, change := range changes {
+				if mdYaml && strings.HasSuffix(change, ".md") ||
+					strings.HasSuffix(change, aliasesFileName) ||
+					strings.HasSuffix(change, ownersFileName) {
+					reusable = false
+					break
+				}
+			}
 		}
+		if reusable {
+			entry.sha = sha
+		} else {
+			if err := gitRepo.Checkout(base); err != nil {
+				return nil, err
+			}
 
-		blacklistConfig := c.config.OwnersDirBlacklist
+			if entry.aliases == nil || entry.sha != sha {
+				// aliases must be loaded
+				entry.aliases = loadAliasesFrom(gitRepo.Dir, log)
+			}
 
-		dirBlacklist := defaultDirBlacklist.Union(sets.NewString(blacklistConfig.Default...))
-		if bl, ok := blacklistConfig.Repos[org]; ok {
-			dirBlacklist.Insert(bl...)
-		}
-		if bl, ok := blacklistConfig.Repos[org+"/"+repo]; ok {
-			dirBlacklist.Insert(bl...)
-		}
-		entry.owners, err = loadOwnersFrom(gitRepo.Dir, mdYaml, entry.aliases, dirBlacklist, log)
-		if err != nil {
-			return nil, fmt.Errorf("failed to load RepoOwners for %s: %v", fullName, err)
+			blacklistConfig := c.ownersDirBlacklist()
+
+			dirBlacklist := defaultDirBlacklist.Union(sets.NewString(blacklistConfig.Default...))
+			if bl, ok := blacklistConfig.Repos[org]; ok {
```

Comment:
Can we put the "get matching blacklists" function into the config package? Look at ConfigAgent.getPlugins() or similar